### DISCUSSION TITLE
puppet-lint and rubocop checks

### DISCUSSION THREAD

**Discussion by sanfrancrisko** [puppetlabs/pdk#1140] [Feature Request]
Originally raised: https://tickets.puppetlabs.com/browse/PDK-1700

We should add/create several puppet-lint plugins to the default install. We should also maintain a set of Puppet specific rubocop rules like https://github.com/chef/cookstyle/tree/master/lib/rubocop/cop/chef

Lint checks:
* puppet-lint-param-docs
* top_scope_facts
* legacy_facts
* world_writable_files
* puppet-lint-security-plugins (author is open to us taking over maintenance)
* https://github.com/TQRG/puppet-security-linter (evaluate first)

Rubocops
* 3.x functions
* facts with code outside of {{setcode}} scope
* facts with dots in their names
* use require_relative for helper code
* don't require parts of puppet that conflict with the autoloader, eg "require 'puppet/type/package'"

**Comment by sanfrancrisko** [comment]
Comment from @genebean :

FWIW, puppet-lint plugin wise, I have traditionally used these:

puppet-lint-absolute_template_path
puppet-lint-classes_and_types_beginning_with_digits-check
puppet-lint-empty_string-check
puppet-lint-leading_zero-check
puppet-lint-legacy_facts-check
puppet-lint-resource_reference_syntax
puppet-lint-spaceship_operator_without_tag-check
puppet-lint-top_scope_facts-check
puppet-lint-trailing_comma-check
puppet-lint-trailing_newline-check
puppet-lint-undef_in_function-check
puppet-lint-unquoted_string-check
puppet-lint-variable_contains_upcase
puppet-lint-version_comparison-check

**Comment by sanfrancrisko** [comment]
Comment from  Tim Muesel:

<p style="margin: 0px; padding: 0px; color: rgb(51, 51, 51); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(244, 245, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">On Vox Pupuli we use:</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(51, 51, 51); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(244, 245, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></p><div id="syntaxplugin" class="syntaxplugin" style="margin: 0px; padding: 0px; color: rgb(51, 51, 51); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(244, 245, 247); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; border: 1px dashed rgb(187, 187, 187); overflow: auto; max-height: 30em; border-radius: 5px !important;">

$ bundle list \| grep puppet-lint-
--
* puppet-lint-absolute_classname-check (3.0.0)
* puppet-lint-anchor-check (1.0.0)
* puppet-lint-classes_and_types_beginning_with_digits-check (0.1.2)
* puppet-lint-file_ensure-check (0.3.1)
* puppet-lint-leading_zero-check (0.1.1)
* puppet-lint-legacy_facts-check (1.0.4)
* puppet-lint-manifest_whitespace-check (0.1.17)
* puppet-lint-param-docs (1.6.0)
* puppet-lint-resource_reference_syntax (1.1.0)
* puppet-lint-strict_indent-check (2.0.7)
* puppet-lint-topscope-variable-check (1.0.1)
* puppet-lint-trailing_comma-check (0.4.2)
* puppet-lint-unquoted_string-check (2.0.0)
* puppet-lint-variable_contains_upcase (1.2.0)
* puppet-lint-version_comparison-check (0.2.1)

</div>

**Comment by sanfrancrisko** [comment]
Comment from @yachub :

Even though options in .puppet-lint.rc are configurable in .sync.yml through Rakefile.linter_fail_on_warnings, I think enabling --fail-on-warnings as a default via changing fail_on_warnings would be a good idea, otherwise any findings caught by the puppet-lint gems results in a successful test. Specifically thinking about CI with pull requests in regards to this.

**Comment by bastelfreak** [comment]
Vox Pupuli now also uses puppet-lint-optional_default-check on all modules. Modules with full puppet-strings coverage also have puppet-lint-param-docs.

**Comment by bastelfreak** [comment]
@sanfrancrisko how can we move this forward? Vox Pupuli publishes a gem that pulls in all related linter plugins: https://rubygems.org/gems/voxpupuli-puppet-lint-plugins  and I think it would be really good for the puppet ecosystem if pdk uses this plugin as well (in addition to setting fail_on_warnings).

**Reply by claviola** [reply]
> @sanfrancrisko how can we move this forward? Vox Pupuli publishes a gem that pulls in all related linter plugins: https://rubygems.org/gems/voxpupuli-puppet-lint-plugins and I think it would be really good for the puppet ecosystem if pdk uses this plugin as well (in addition to setting fail_on_warnings).

It looks like this is the case now: https://github.com/puppetlabs/pdk-templates/commit/6bf31242
